### PR TITLE
feat: require string tokens used as comparison values to be delimited

### PR DESCRIPTION
**Pull request by jimlambrt**
Originally, you could could write a query like: `name=alice`

This PR requires that same query to be written as: `name="alice"`

This change will allow us to introduce new keywords and/or reserved words as needed in the future without introducing breaking changes.  It also disambiguates symbols from strings.

Also, this PR sets us up to support additional string delimiters in a subsequent PR.  For instance, we can support `` ` `` (backtick) and `'` (singe quote) as string delimiters which would really improve the UX for writing queries. 

This PR also includes: 

- [feat: add support to escape double quotes](https://github.com/hashicorp/mql/commit/1e976760cc7b8b91ee3158abd88a271681ecba53)
- [refact: split lexStringState() into two distinct state functions](https://github.com/hashicorp/mql/pull/23/commits/e13f285bc36df22b90847a3d5361c9657eec6204)
- [refact (lexSymbolState): eliminate the bytes.Buffer](https://github.com/hashicorp/mql/pull/23/commits/48d1e235258ad38962bd53995dbeaa1feebb091b)

### CODE REVIEW

**Comment by jimlambrt** [comment]
I still need to update the docs and add a CHANGELOG entry.  Besides those chores, this PR is ready for review.

**Comment by jefferai** [comment]
> Also, this PR sets us up to support additional string delimiters in a subsequent PR. For instance, we can support ` (backtick) and ' (singe quote) as string delimiters which would really improve the UX for writing queries.

Do we want this? Would it be better to standardize on a single delimiter so that you can include others without escaping? e.g. if we standardize on single quote, we don't need to escape embedded double quotes. Or if we standardize on double quotes we don't need to escape apostrophes.

**Comment by jimlambrt** [comment]
The idea is to make it easier for the user by allowing them to choose the most convenient delimiter for the query they're writing.

Examples using double-quotes, single-quotes and backticks:

- description=`Alice's Restaurant` 
- description="Alice's Restaurant" 
- description = 'This restaurant is not "owned" by Alice'

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ application to SQL injection.
 ### [github.com/go-gorm/gorm](https://github.com/go-gorm/gorm)
 
 ```Go
-w, err := mql.Parse("name=alice or name=bob)",User{})
+w, err := mql.Parse(`name="alice" or name="bob")`,User{})
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply README.md:17]: ty!

File: lex.go [hunk 2]
```diff
@@ -77,65 +83,144 @@ func lexStartState(l *lexer) (lexStateFunc, error) {
...
+			break ReadRunes
+		default:
+			continue ReadRunes
 		}
 	}
 
 	// before emitting a token, do we have a special string?  But, first let's
 	// check if we're dealing with a quoted string, since we want to support
 	// emitting string tokens for "and", "or" so those tokens can be used in
 	// comparison expr. Example: name % "Johnson and"
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply lex.go:180]: deleted. ty!

File: GRAMMAR.md [hunk 3]
```diff
@@ -40,15 +40,17 @@ A `condition` is any expression that evaluates to a result of type boolean.
...
 
 ### string
 
 * \<string>
 
 ### quoted string
 
-A string delimited by quotes. Note: you can use a quoted string whenever you
-need to include logical operators or whitespace in a string value.
+A string delimited by quotes. Supported delimiters: double-quotes
```
Comment (johanbrandhorst) [inline GRAMMAR.md:53]: Mind adding that one can escape double quotes within a quoted string with a backslash? And that a backslash can be escaped with another backslash?

**Review by johanbrandhorst — APPROVED** [review]
Could you add some quoted strings to the fuzz tests?

File: GRAMMAR.md (hunk 3 above)
Comment (jimlambrt) [reply GRAMMAR.md:53]: ty!

**Pull request merged**